### DISCUSSION TITLE
Webpack 5 Web Fonts not loading??

### DISCUSSION THREAD

**Discussion by devhoussam** [webpack/webpack#13269] [Q&A]
I'm using webpack 5 the latest version of webpack it brings many cool features but unfortunately, I got a lot of issues while I trying to load web fonts. I install both of font-awesome 5 and bootstrap icons npm package in my project and include it inside my `app.scss` when I run `npm run build` command to create a dist folder a got this error.

![Image](https://i.stack.imgur.com/ABMkp.png)

![Image](https://i.stack.imgur.com/vGMNR.png)

**webpack.config.js**

```
const path = require("path");
const webpack = require("webpack");
const HtmlWebpackPlugin = require("html-webpack-plugin");
const TerserPlugin = require("terser-webpack-plugin");
const MiniCssExtractPlugin = require("mini-css-extract-plugin");

module.exports = {
    mode: process.env.NODE_ENV,
    entry: {
        app: [path.resolve(__dirname, "src/js/app.js"), path.resolve(__dirname, "src/scss/app.scss")],
    },
    output: {
        path: path.resolve(__dirname, "dist"),
        filename: "js/[name].bundle.js",
        assetModuleFilename: "images/[hash][ext][query]",
        clean: true,
    },
    devServer: {
        contentBase: path.join(__dirname, "dist"),
        compress: true,
        open: true,
        port: 3000,
    },
    devtool: "source-map",
    module: {
        rules: [
            {
                test: /\.js$/,
                exclude: /(node_modules)/,
                use: {
                    loader: "babel-loader",
                    options: {
                        presets: ["@babel/preset-env"],
                        plugins: ["@babel/plugin-proposal-class-properties"],
                        cacheDirectory: true,
                    },
                },
            },
            {
                test: /\.s[ac]ss$/i,
                use: [
                    process.env.NODE_ENV !== "production" ? "style-loader" : MiniCssExtractPlugin.loader,
                    {
                        loader: "css-loader",
                        options: {
                            sourceMap: true,
                        },
                    },
                    {
                        loader: "postcss-loader",
                        options: {
                            implementation: require("postcss"),
                            sourceMap: true,
                            postcssOptions: {
                                plugins: [["postcss-preset-env", { browsers: "last 2 versions" }]],
                            },
                        },
                    },
                    {
                        loader: "sass-loader",
                        options: {
                            implementation: require("sass"),
                            sourceMap: true,
                            sassOptions: {
                                fiber: require("fibers"),
                                outputStyle: "compressed",
                            },
                        },
                    },
                ],
            },
            {
                test: /\.(ico|png|svg|jpg|jpeg|gif|svg|webp|tiff)$/i,
                type: "asset/resource",
                generator: {
                    filename: "images/[name].[hash][ext]",
                },
            },
            {
                test: /\.(svg|woff|woff2|eot|ttf|otf)$/i,
                type: "asset/inline",
            },
        ],
    },
    plugins: [
        new HtmlWebpackPlugin({
            template: path.resolve(__dirname, "src/index.html"),
            inject: "body",
        }),
        new MiniCssExtractPlugin({
            filename: "css/[name].bundle.css",
            chunkFilename: "css/[id].bundle.css",
        }),
    ],
    optimization: {
        minimize: true,
        minimizer: [new TerserPlugin()],
        splitChunks: {
            chunks: "all",
        },
        runtimeChunk: {
            name: "runtime",
        },
    },
    performance: {
        hints: false,
    },
    experiments: {
        asset: true,
    },
};
```

**src/scss/app.scss**

```
@import "~bootstrap-icons/font/bootstrap-icons";

$fa-font-path: "@fortawesome/fontawesome-free/webfonts";

@import "~@fortawesome/fontawesome-free/scss/fontawesome";
@import "~@fortawesome/fontawesome-free/scss/solid";
@import "~@fortawesome/fontawesome-free/scss/brands";
@import "~@fortawesome/fontawesome-free/scss/regular";
```

I try to find how to fix this issue by locking for other example but I didn't find anything useful.

**Comment by alexander-akait** [comment]
You have invalid `url()` in SCSS, because sass (i.e. node-sass/dart-sass) compile all sass/scss files in one by entry path (place where you import sass file in js or in entrypoint) you need do `url()` in nested files as relative to entry file

**Reply by devhoussam** [reply]
> You have invalid `url()` in SCSS, because sass (i.e. node-sass/dart-sass) compile all sass/scss files in one by entry path (place where you import sass file in js or in entrypoint) you need do `url()` in nested files as relative to entry file

like this ?

```
@import url("~bootstrap-icons/font/bootstrap-icons");

$fa-font-path: "@fortawesome/fontawesome-free/webfonts";

@import url("~@fortawesome/fontawesome-free/scss/fontawesome");
@import url("~@fortawesome/fontawesome-free/scss/solid");
@import url("~@fortawesome/fontawesome-free/scss/brands");
@import url("~@fortawesome/fontawesome-free/scss/regular");
```

**Reply by alexander-akait** [reply]
I don't know your development paths, so hard to say, where locales fonts and where you import entry sass file?

**Reply by devhoussam** [reply]
> I don't know your development paths, so hard to say, where locales fonts and where you import entry sass file?

I made this repository https://github.com/devhoussam/webpack-5-boilerplate-2021/tree/master

**Reply by alexander-akait** [reply]
You have:
```css
@font-face {
  font-family: "bootstrap-icons";
  src: url("./fonts/bootstrap-icons.woff2?231ce25e89ab5804f9a6c427b8d325c9") format("woff2"),
url("./fonts/bootstrap-icons.woff?231ce25e89ab5804f9a6c427b8d325c9") format("woff");
}
```

in `@import "~bootstrap-icons/font/bootstrap-icons";`, so you need to use `resolve.alias`, i.e.

```js
module.exports = {
  // Other options
  resolve: {
        alias: {
            './fonts/bootstrap-icons.woff': path.resolve(__dirname, './node_modules/bootstrap-icons/font/fonts/bootstrap-icons.woff'),
            './fonts/bootstrap-icons.woff2': path.resolve(__dirname, './node_modules/bootstrap-icons/font/fonts/bootstrap-icons.woff2'),
        }
  },
};
```

Also you have invalid configuration for assets, please use:
```
{
  test: /\.(ico|png|jpg|jpeg|gif|svg|webp|tiff)$/i,
  type: "asset/resource",
  generator: {
    filename: "images/[name].[hash][ext]",
  },
},
{
  test: /\.(woff|woff2|eot|ttf|otf)$/i,
  type: "asset/inline",
},
```

Avoid overlapping extensions in `test`

**Reply by devhoussam** [reply]
> You have:
> 
> ```css
> @font-face {
>   font-family: "bootstrap-icons";
>   src: url("./fonts/bootstrap-icons.woff2?231ce25e89ab5804f9a6c427b8d325c9") format("woff2"),
> url("./fonts/bootstrap-icons.woff?231ce25e89ab5804f9a6c427b8d325c9") format("woff");
> }
> ```
> 
> in `@import "~bootstrap-icons/font/bootstrap-icons";`, so you need to use `resolve.alias`, i.e.
> 
> ```js
> module.exports = {
>   // Other options
>   resolve: {
>         alias: {
>             './fonts/bootstrap-icons.woff': path.resolve(__dirname, './node_modules/bootstrap-icons/font/fonts/bootstrap-icons.woff'),
>             './fonts/bootstrap-icons.woff2': path.resolve(__dirname, './node_modules/bootstrap-icons/font/fonts/bootstrap-icons.woff2'),
>         }
>   },
> };
> ```
> 
> Also you have invalid configuration for assets, please use:
> 
> ```
> {
>   test: /\.(ico|png|jpg|jpeg|gif|svg|webp|tiff)$/i,
>   type: "asset/resource",
>   generator: {
>     filename: "images/[name].[hash][ext]",
>   },
> },
> {
>   test: /\.(woff|woff2|eot|ttf|otf)$/i,
>   type: "asset/inline",
> },
> ```
> 
> Avoid overlapping extensions in `test`

and what about **Font Awesome ?**

**Reply by alexander-akait** [reply]
All fine

**Comment by GarikGelios** [comment]
@alexander-akait thanks for your help and time!
Unfortunately I didn't succeed with your example. The console gives an error and there is no folder or file with fonts in the _dist/_

```
bundle.css:1          GET http://127.0.0.1:5500/dist/fonts/bootstrap-icons.woff2?24e3eb84d0bcaf83d77f904c78ac1f47 404 (Not Found)
bundle.css:1          GET http://127.0.0.1:5500/dist/fonts/bootstrap-icons.woff?24e3eb84d0bcaf83d77f904c78ac1f47 net::ERR_ABORTED 404 (Not Found)
```
my **webpack,config.js** file contain _resolve.alias_ as you advised

```
const path = require('path')

module.exports = {
  entry: './assets/javascript/dev-main.js', 
  resolve: {
    alias: {
      './fonts/bootstrap-icons.woff': path.resolve(__dirname, './node_modules/bootstrap-icons/font/fonts/bootstrap-icons.woff'),
      './fonts/bootstrap-icons.woff2': path.resolve(__dirname, './node_modules/bootstrap-icons/font/fonts/bootstrap-icons.woff2')
    } // nothing has changed in the bundle file, the path remains the same
  },
  output: {
    clean: true, // Clean the output directory before emit.
    filename: 'main.js',
    path: path.resolve(__dirname, 'dist')
  },
  mode: 'development',
  module: {
    rules: [
      {
        test: /\.js$/,
        use: []
      },
      {
        test: /\.scss$/,
        type: 'asset/resource',
        generator: {
          filename: 'bundle.css'
        },
        use: ['sass-loader']
      },
      {
        test: /\.(woff|woff2|eot|ttf|otf)$/i, // fonts not working
        type: 'asset/inline'  // I tried all combinations of loaders and the old approach and specifying the folder and the name of the output file - there was no success =(
      },
      {
        mimetype: 'image/svg+xml', // icons don't work either
        scheme: 'data',
        type: 'asset/resource', // I added icons in all ways, they appeared if the entire SVG was inserted, but there were no changes in the dist/ 
        generator: {
          filename: 'icons/[hash].svg'
        }
      }
    ]
  }
}

```
my **dependencies** from package.json file

```
"devDependencies": {
    "@popperjs/core": "^2.11.6",
    "bootstrap": "^5.2.3",
    "css-loader": "^6.7.2",
    "sass": "^1.56.1",
    "sass-loader": "^13.2.0",
    "standard": "^17.0.0",
    "style-loader": "^3.3.1",
    "stylelint-config-standard-scss": "^6.1.0",
    "webpack": "^5.75.0",
    "webpack-cli": "^5.0.0"
  },
  "dependencies": {
    "bootstrap-icons": "^1.10.2",
    "tiny-slider": "^2.9.4"
  }
```
Bootstrap styles and icons are included in the style file **main.scss**
```
// Bootstrap 5
@import "~bootstrap/scss/bootstrap";
@import "~bootstrap-icons/font/bootstrap-icons";
```
also in **dev-main.js**
```
// Import our custom SCSS
import '../stylesheets/sass/main.scss'

// Import all of Bootstrap's JS
import 'bootstrap'
```
Styles and Javascript work fine, but there are no icons. Even the font added to the source files does not appear in the _dist/_, although it is connected (_in main.scss_) and used.
```
@font-face {
  font-family: Roboto;
  src: url("../../fonts/Roboto-Medium.ttf") format("ttf");
}

.font-roboto {
  font-family: Roboto, sans-serif;
}
```
my **assets** folder
```
|-fonts/
|  |-Roboto-Medium.ttf
|-img/
|-javascript/
|  |-dev-main.js
|-stylesheets/sass/
|  |-main.scss
```
after `npm run build` (_in my package.json is_ `"build": "webpack --watch"`) I get only two files in my _dist/_
```
|-dist/
|  |-bundle.css
|  |-main.js
```
______________________________________________________
There are several solutions for icons, but they are not very good, this is not how it should work.
1. We can connect fonts via CDN, as indicated in the documentation (_both `<link>` and `@import`_), everything will work fine.
2. Specify an explicit path to the fonts in the modules folder:
```
@font-face {
  font-family: bootstrap-icons;
  src: url("./../../../node_modules/bootstrap-icons/font/fonts/bootstrap-icons.woff") format("woff");
}
```
3. Embed all SVG icons into the document. But it won't help with fonts

**Reply by alexander-akait** [reply]
Why do you use:
```
{
        test: /\.scss$/,
        type: 'asset/resource',
        generator: {
          filename: 'bundle.css'
        },
        use: ['sass-loader']
      },
```

Do you want only compile sass and load CSS as a file?

**Reply by GarikGelios** [reply]
For now, that's all I need. I have `@import` all the _.sсss_ files into the _main.sсss_ which webpack turns me into one regular _bundle.сss_ ← which I include it in the my _index.html_ file.
I don't have the need for additional configurations yet, or I don't know their benefits yet. So far, the actual task with fonts. If there are any scss configurations to help me with the fonts, I'd be grateful for the info.

**Reply by alexander-akait** [reply]
@GarikGelios I recommend to use https://github.com/webpack-contrib/sass-loader#1-mini-css-extract-plugin

**Reply by GarikGelios** [reply]
@alexander-akait  Thanks, yes I think I should add file minification and autoprefixer too. But that won't solve the font issue, will it?

**Reply by alexander-akait** [reply]
@GarikGelios you have `type: 'asset/resource'`, you should not use this in your case, so you have the problem with fonts

**Reply by GarikGelios** [reply]
@alexander-akait in `test: /\.scss$/` if I use 
1. `type: asset/inline` generates an error fot console
2. `type: asset/source` does not generate stylesheet
3. `type: asset` automatically chooses between exporting a data URI and emitting a separate file. Everything is the same as `withasset/resource`

Files with fonts did not appear in the destination folder. In the destination folder in the style, all urls are indicated as in the sources.
Bootstrap expects a folder with a file `src: url("./fonts/bootstrap-icons.woff2)`, but it is not transferred from the module folder.
Added fonts to the source folder, saved the same path in the final style file `url("../../fonts/Roboto-Medium.ttf")`.
Files don't migrate and paths don't adapt.

**Reply by alexander-akait** [reply]
You should not to use `type: asset/inline`/`type: asset/source`/`type: asset` for `scss` (`css` or `less`, except some cases when you really need to load your code as file, not as module in bundle), please look at the example and you don't found `type: asset`, if you set `type: asset`, your will load your CSS as file (not as module, just a simple text file), so fonts will not work

**Reply by GarikGelios** [reply]
@alexander-akait  it's a miracle, it worked! Thanks!
I don't really understand why the fonts are not extracted when creating a separate style file.
```
{
  test: /\.s[ac]ss$/i,
  // type: 'asset/resource', // type creates a separate file, which makes it impossible to get the fonts
  generator: {
  filename: 'bundle.css' // now all styles are in javascript, the file is not created
  },
use: [
   'style-loader',  // Creates `style` nodes from JS strings
   'css-loader',// Translates CSS into CommonJS
   'sass-loader' // Compiles Sass to CSS
  ]
},
{
  test: /\.(woff|woff2|eot|ttf|otf)$/i, // fonts finally work!
  type: 'asset/resource',
  generator: {
    outputPath: 'fonts/', // how to force styles to look in this folder?
    filename: '[name].[ext]'
  }
},
```
one issue solved, two new ones appeared)
1. how to split styles and link as a separate file
2. how to force look fonts in _dist/fonts/_ instead _dist/_ folder

**Reply by alexander-akait** [reply]
> how to split styles and link as a separate file

What do you mean? If you use `sass` you will always have one bundled CSS, because sass compile them into one file, you can split your styles into files and use `import "./style.scss";` where you need require style

> how to force look fonts in dist/fonts/ instead dist/ folder

```
{
  test: /\.(woff|woff2|eot|ttf|otf)$/i, // fonts finally work!
  type: 'asset/resource',
  generator: {
    filename: 'fonts/[hash][ext][query]'
  }
}  
```

Also please remove the `generator` option for `test: /\.s[ac]ss$/i,`, you don't need it there

**Reply by GarikGelios** [reply]
@alexander-akait  Thanks! path to fonts via filename worked, I tried before `outputPath: 'fonts/'`
What I mean about split styles:
In the source, I have a lot of style files, all connected to the _main.scss_ via `@import`. When I used  `type: 'asset/resource'` for `test: /\.s[ac]ss$/i` I received _bundle.css_ in _dist/_ which was connected in _index.html_ via `<link rel="stylesheet" href="dist/bundle.css">` — I like this approach with a separate connection of styles and scripts.
Now I have the following structure and everything works, i.e. styles from the javascript file
```
|-dist/
|  |- fonts/
|    |- [hash][ext][query]
|  |- main.js
|  |- bundle.css ← this file is no longer created separately
```

**Reply by alexander-akait** [reply]
> In the source, I have a lot of style files, all connected to the main.scss via @import

In such case you will have the one big CSS files, you need to refactor your structure, we can't split your big CSS files, because we don't know how to do it (i.e. your order and when you need something when is not)

**Reply by GarikGelios** [reply]
@alexander-akait i don't want to split the style file into many other files, i want one big CSS from one big SCSS. I want styles in a separate file from javascript, as it was before:
```
|-dist/
|  |- main.js
|  |- bundle.css
```
and now, what I have after:
```
|-dist/
|  |- fonts/
|    |- [hash][ext][query]
|  |- main.js
```

**Reply by alexander-akait** [reply]
Do you use my recommendation for a configuration? If you will have this for `production` build, for development it is faster to have in JS, just run `npx webpack --mode production`

**Reply by GarikGelios** [reply]
Thanks to your recommendations, everything works locally for development, but I want includ into the html header a separate stylesheet file.
I agree that for development it is faster to have style in JS (_I believe you as a professional, I do not see the difference_).
If I run `npx webpack --mode production` or set `mode: 'production'` in _webpack.config.js_ — main.css file does not appear in dist folder.

**Reply by alexander-akait** [reply]
check you have `MiniCssExtractPlugin` in your configuratio as in my reccomendation https://github.com/webpack-contrib/sass-loader#1-mini-css-extract-plugin, it will extract CSS for production usage

**Reply by GarikGelios** [reply]
I installed and added **MiniCssExtractPlugin** , but in prodaction mode it give me only _main.js.LICENSE.txt_

```
const MiniCssExtractPlugin = require('mini-css-extract-plugin')
const path = require('path')

module.exports = {
  entry: './assets/javascript/dev-main.js', // everything will merge into a single main.js. If you want separate files  lets create an object. Read more at concepts/entry-points/
  resolve: {
    alias: {
      './fonts/bootstrap-icons.woff': path.resolve(__dirname, './node_modules/bootstrap-icons/font/fonts/bootstrap-icons.woff'),
      './fonts/bootstrap-icons.woff2': path.resolve(__dirname, './node_modules/bootstrap-icons/font/fonts/bootstrap-icons.woff2')
    }
  },
  output: {
    clean: true, // Clean the output directory before emit.
    filename: 'main.js',
    path: path.resolve(__dirname, 'dist')
  },
  mode: 'production',
  module: {
    rules: [
      {
        test: /\.js$/,
        use: []
      },
      {
        test: /\.s[ac]ss$/i,
        use: [
          // fallback to style-loader in development
          process.env.NODE_ENV !== 'production' // check for which mode we work
            ? 'style-loader' // Creates `style` nodes from JS strings, for development it is faster to have style in JS
            : MiniCssExtractPlugin.loader,
          'css-loader', // Translates CSS into CommonJS
          'sass-loader' // Compiles Sass to CSS
        ]
      },
      {
        test: /\.(woff|woff2|eot|ttf|otf)$/i,
        type: 'asset/resource',
        generator: {
          filename: 'fonts/[hash][ext][query]'
        }
      },
      {
        test: /\.(ico|png|jpg|jpeg|gif|svg|webp|tiff)$/i,
        type: 'asset/resource',
        generator: {
          filename: 'images/[name].[hash][ext]'
        }
      }
    ]
  },
  plugins: [
    new MiniCssExtractPlugin({
      filename: '[name].css',
      chunkFilename: '[id].css'
    })
  ]
}

```

**Reply by alexander-akait** [reply]
You have `process.env.NODE_ENV`, so please use `npx webpack --node-env production`

**Reply by GarikGelios** [reply]
I just copy this `process.env.NODE_ENV !== 'production' ? 'style-loader'` from https://github.com/webpack-contrib/sass-loader#1-mini-css-extract-plugin.
When I run `npx webpack --node-env production` I get:
```
[webpack-cli] Error: Unknown option '--node-env'
[webpack-cli] Run 'webpack --help' to see available commands and options
```

**Reply by alexander-akait** [reply]
Please update webpack-cli to the latest stable version
